File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

 /*
 1. let good = 0
 2. dfs (node, max) {
    if (!node) return
    if (node.val >= max) {
        good++
    }
    dfs(node.left, Math.max(max, node.max))
    dfs(node.right, Math.max(max, node.max))
    }
3. dfs(root, root.val)
4. return good
 */

var goodNodes = function(root) {
    var good = 0

    var dfs = (node, max) => {
        if (!node) {
            return
        }
        if (node.val >= max) {
            good++
        }
        dfs(node.left, Math.max(max, node.val))
        dfs(node.right, Math.max(max, node.val))
    }
    dfs(root, root.val)
    return good
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = 0

        def dfs(root, maxSoFar):
            nonlocal res

            if not root:
                return
            
            if root.val >= maxSoFar:
                maxSoFar = root.val
                res += 1
            
            if root.left:
                dfs(root.left, maxSoFar)
            if root.right:
                dfs(root.right, maxSoFar)

        dfs(root, root.val)
        return res